tf2onnx/rewriter/gemm_rewriter.py

```diff
@@ -0,0 +1,118 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT license.
+
+"""
+tf2onnx.rewrite - rewrite tensorflow subgraph to onnx gemm op
+"""
+import logging
+from onnx import onnx_pb
+from tf2onnx.graph_matcher import OpTypePattern, GraphMatcher
+
+# pylint: disable=missing-docstring
+
+def rewrite_gemm(g, ops):
+    if g.opset <= 6:
+        return ops
+
+    # pattern0: alpha*A*B + beta*C
+    pattern0 = \
+        OpTypePattern('Add', name='add', inputs=[
+            OpTypePattern('Mul', name='mul1', inputs=[
+                OpTypePattern('Const', name='alpha'),
+                OpTypePattern('MatMul', name='matmul')
+            ]),
+            OpTypePattern('Mul', name='mul2', inputs=[
+                OpTypePattern('Const', name='beta'),
+                OpTypePattern('*', name='C')
+            ])
+        ])
+
+    # pattern1: alpha*A*B + C
+    pattern1 = \
+        OpTypePattern('Add', name='add', inputs=[
+            OpTypePattern('Mul', name='mul1', inputs=[
+                OpTypePattern('MatMul', name='matmul'),
+                OpTypePattern('Const', name='alpha')
+            ]),
+            OpTypePattern('*', name='C'),
+        ])
+
+    # pattern2: A*B + beta*C
+    pattern2 = \
+        OpTypePattern('Add', name='add', inputs=[
+            OpTypePattern('MatMul', name='matmul'),
+            OpTypePattern('Mul', name='mul2', inputs=[
+                OpTypePattern('Const', name='beta'),
+                OpTypePattern('*', name='C')
+            ])
+        ])
+
+    # pattern3: A*B + C
+    pattern3 = \
+        OpTypePattern('Add', name='add', inputs=[
+            OpTypePattern('MatMul', name='matmul'),
+            OpTypePattern('*', name='C'),
+        ])
+
+    pattern_list = [pattern0, pattern1, pattern2, pattern3]
+
+    for pattern_id, pattern in enumerate(pattern_list):
+        matcher = GraphMatcher(pattern, allow_reorder=True)
+        match_results = list(matcher.match_ops(ops))
+        if match_results:
+            for match in match_results:
+                add_node = match.get_op('add')
+                matmul_node = match.get_op("matmul")
+                input_c_node = match.get_op("C")
+
+                if g.get_dtype(matmul_node.input[0]) != onnx_pb.TensorProto.FLOAT:
+                    logging.warning(u"For now, onnxruntime only support float type for Gemm rewriter")
+                    return ops
+                a_edge_name = matmul_node.input[0]
+                b_edge_name = matmul_node.input[1]
+                c_edge_name = input_c_node.output[0]
+                attr = {}
+
+                # For each pattern, we must ensure that alpha and beta are both scalar, or return ops
+                if pattern_id == 0:  # pattern 0: alpha*A*B + beta*C
+                    alpha = match.get_op("alpha").get_tensor_value()
+                    beta = match.get_op("beta").get_tensor_value()
+                    if isinstance(alpha, (float, int)):
```

Comment:
can they be different dtype: add, mul1, mul2, alpha, beta?

Reply:
Thanks for your comment. I have improved these code, please have a review on them.